idaes/core/util/tables.py

```diff
@@ -266,22 +268,50 @@ def stream_table_dataframe_to_string(stream_table, **kwargs):
         na_rep=na_rep, justify=justify, float_format=float_format, **kwargs
     )
 
-
-def _get_state_from_port(port, time_point):
-    # Check port for _state_block attribute
-    try:
-        if len(port._state_block) == 1:
-            return port._state_block[0][time_point]
-        else:
-            return port._state_block[0][time_point, port._state_block[1]]
-    except AttributeError:
-        # Port was not created by IDAES add_port methods. Return exception for
-        # the user to fix.
-        raise ConfigurationError(
-            f"Port {port.name} does not have a _state_block attribute, "
-            f"thus cannot determine StateBlock to use for collecting data."
-            f" Please provide the associated StateBlock instead, or use "
-            f"the IDAES add_port methods to create the Port."
+def _get_state_from_port(port,time_point):
+    from idaes.core.property_base import StateBlockData
+    states = list()
+    v = None # To make PyLint stop complaining
+    for v in port.iter_vars():
+        if not v.parent_block().parent_component().is_indexed():
+            raise TypeError(
+                f"No state block could be retrieved from Port {port.name} "
+                f"because variable {v.name} does not belong to a state block."
+                )
+        if not isinstance(v.parent_block(),StateBlockData):
+            raise TypeError(
```

Comment:
Do we really need to insist upon state blocks here as well? For a general purpose method, would it not be sufficient to check to see that all members of the Port simply have the same parent component and return that if true? I.e. At this stage do we actually care what the parent component is and whether it is indexed?

Later tools might care, but it should be up to them to make that decision.

Reply:
I think I agree with @andrewlee94 here.  I could imagine adding a block to mimic a state block, but is not a state block.  We could leave it to other things down the line to complain if some expected thing is missing.

Reply:
In principle, I would like to support duck typing. However, I'm not sure whether that's worth the effort to support at this stage. Previously, the function would work only for `Port`s on `UnitModelBlock`s, so no functionality is being lost by type checking here.  Right now, the functionality is being used only to create stream tables, for which the function is arguably overkill (I think we could just read the variables/expressions directly from the port). For my global balance functionality, I might use this function during prototyping, but there's a significant chance that functionality to get the relevant info is added on some other object.

Reply:
Type checking has been removed at this point.